File: src/scraper.ts
import { ApifyClient } from 'apify-client';
import { Config } from './config';

export interface SearchParams {
    country?: string;
    state?: string;
    city?: string;
    area?: string;
    category?: string;
    maxResults?: number;
}

export class GoogleMapsScraper {
    private client: ApifyClient;

    constructor() {
        if (!Config.APIFY_API_TOKEN) {
            throw new Error("APIFY_API_TOKEN is not set in environment variables.");
        }
        this.client = new ApifyClient({ token: Config.APIFY_API_TOKEN });
    }

    async startSearch({ country, state, city, area, category, maxResults = 20 }: SearchParams) {
        const searchParts: string[] = [];
        if (category) searchParts.push(category);
        if (area) searchParts.push(area);
        if (city) searchParts.push(city);
        if (state) searchParts.push(state);
        if (country) searchParts.push(country);

        const searchQuery = searchParts.join(", ");
        if (!searchQuery) {
            throw new Error("At least one search parameter must be provided.");
        }

        console.log(`Starting Async Apify Scraper for query: '${searchQuery}' (max ${maxResults} results)`);

        const runInput = {
            searchStringsArray: [searchQuery],
            maxCrawledPlacesPerSearch: maxResults,
            language: "en",
            maxImages: 0,
            maxReviews: 0,
            scrapeReviewerName: false,
            scrapeReviewerId: false,
            scrapeReviewerUrl: false,
            scrapeResponseFromOwnerText: false,
        };

        const run = await this.client.actor(Config.APIFY_ACTOR_ID).start(runInput);
        console.log(`Scraper started. Run ID: ${run.id}`);
        
        return run;
    }

    async checkRunStatus(runId: string) {
        return await this.client.run(runId).get();
    }

    async getDatasetItems(datasetId: string) {
        const { items } = await this.client.dataset(datasetId).listItems();
        return items;
    }
}

File: scraper.py
from apify_client import ApifyClient
from config import Config

class GoogleMapsScraper:
    def __init__(self):
        if not Config.APIFY_API_TOKEN:
            raise ValueError("APIFY_API_TOKEN is not set in environment variables.")
        self.client = ApifyClient(Config.APIFY_API_TOKEN)
        
    def search_businesses(self, country=None, state=None, city=None, area=None, category=None, max_results=20):
        """
        Builds the search string based on the provided parameters and runs the scraper.
        """
        search_parts = []
        if category:
            search_parts.append(category)
        if area:
            search_parts.append(area)
        if city:
            search_parts.append(city)
        if state:
            search_parts.append(state)
        if country:
            search_parts.append(country)
            
        search_query = ", ".join(search_parts)
        if not search_query:
            raise ValueError("At least one search parameter must be provided.")
            
        print(f"Starting Apify Scraper for query: '{search_query}' (max {max_results} results)")
        
        # Prepare the Actor input
        run_input = {
            "searchStringsArray": [search_query],
            "maxCrawledPlacesPerSearch": max_results,
            "language": "en",
            "maxImages": 0,
            "maxReviews": 0,
            "scrapeReviewerName": False,
            "scrapeReviewerId": False,
            "scrapeReviewerUrl": False,
            "scrapeResponseFromOwnerText": False,
        }
        
        # Run the Actor and wait for it to finish
        run = self.client.actor(Config.APIFY_ACTOR_ID).call(run_input=run_input)
        
        print(f"Scraper finished. Run ID: {run['id']}")
        print("Fetching results...")
        
        # Fetch and yield the results from the run's dataset
        dataset_items = self.client.dataset(run["defaultDatasetId"]).iterate_items()
        
        results = list(dataset_items)
        print(f"Fetched {len(results)} raw items from Apify.")
        return results
